src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+        quota_info.max_objects = std::stoull(it->second.c_str());
+	attrs.erase(it->first);
+      } else if (it->first == "max_buckets") {
+        source->get_bucket()->get_owner()->set_max_buckets(std::stoull(it->second.c_str()));
+	attrs.erase(it->first);
       }
     }
+    }
 
     source->get_bucket()->get_owner()->set_info(quota_info);
     source->set_obj_state(*astate);
    
     /* Set attributes locally */
-    int setAttrsReturn = source->set_attrs(newAttrs);
+    int set_attrsReturn = source->set_attrs(attrs);
 
-    if (setAttrsReturn < 0) {
-      ldpp_dout(dpp, 20) << "D4N Filter: Cache get object operation failed." << dendl;
+    if (set_attrsReturn < 0) {
+      ldpp_dout(dpp, 20) << "D4N Filter: Cache set object operation failed." << dendl;
     } else {
-      ldpp_dout(dpp, 20) << "D4N Filter: Cache get object operation succeeded." << dendl;
+      ldpp_dout(dpp, 20) << "D4N Filter: Cache set object operation succeeded." << dendl;
     }   
   }
 
   return ret;
 }
 
+void D4NFilterObject::D4NFilterReadOp::cancel() {
+  aio->drain();
+}
+
+int D4NFilterObject::D4NFilterReadOp::drain(const DoutPrefixProvider* dpp) {
+  auto c = aio->wait();
+  while (!c.empty()) {
+    int r = flush(dpp, std::move(c));
+    if (r < 0) {
+      cancel();
+      return r;
+    }
+    c = aio->wait();
+  }
+  return flush(dpp, std::move(c));
+}
+
+int D4NFilterObject::D4NFilterReadOp::flush(const DoutPrefixProvider* dpp, rgw::AioResultList&& results) {
+  int r = rgw::check_for_errors(results);
+
+  if (r < 0) {
+    return r;
+  }
+
+  std::list<bufferlist> bl_list;
+
+  auto cmp = [](const auto& lhs, const auto& rhs) { return lhs.id < rhs.id; };
+  results.sort(cmp); // merge() requires results to be sorted first
+  completed.merge(results, cmp); // merge results in sorted order
+
+  ldpp_dout(dpp, 20) << "D4NFilterObject::In flush:: " << dendl;
+
+  while (!completed.empty() && completed.front().id == offset) {
+    auto bl = std::move(completed.front().data);
+
+    ldpp_dout(dpp, 20) << "D4NFilterObject::flush:: calling handle_data for offset: " << offset << " bufferlist length: " << bl.length() << dendl;
+
+    bl_list.push_back(bl);
+    offset += bl.length();
+    int r = client_cb->handle_data(bl, 0, bl.length());
+    if (r < 0) {
+      return r;
+    }
+    completed.pop_front_and_dispose(std::default_delete<rgw::AioResultEntry>{});
+  }
+
+  ldpp_dout(dpp, 20) << "D4NFilterObject::returning from flush:: " << dendl;
+  return 0;
+}
+
+int D4NFilterObject::D4NFilterReadOp::iterate(const DoutPrefixProvider* dpp, int64_t ofs, int64_t end,
+                        RGWGetDataCB* cb, optional_yield y) 
+{
+  const uint64_t window_size = g_conf()->rgw_get_obj_window_size;
+  std::string oid = source->get_key().get_oid();
+
+  ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << "oid: " << oid << " ofs: " << ofs << " end: " << end << dendl;
+
+  this->client_cb = cb;
+  this->cb->set_client_cb(cb, dpp); // what's this for? -Sam
```

Comment:
What is the difference between `D4NFilterReadOp::client_cb` and `D4NFilterGetCB::cb`?

Reply:
There are two callbacks here:
1. One the iterate method receives from the upper layer to send back data read.
2. The second callback belongs to the Filterdriver.
Both are explained below:
`cb` is the callback that we receive from the upper layer. You can see https://github.com/ceph/ceph/blob/main/src/rgw/rgw_op.cc#L2205 and the callback is being passed to the `iterate` method here: https://github.com/ceph/ceph/blob/main/src/rgw/rgw_op.cc#L2416. `cb` is being saved in `this->client_cb` in `D4NFilterObject::D4NFilterReadOp`. The callback method of the filter driver is `D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data`, which is stored in `this->cb`. `this->cb` is passed to the backend store, so that when data is sent back from the store, the filter driver callback  intercepts it and writes it to the cache (as you can see in `D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data`). Now the same data needs to be sent back to the upper layer/client also, so the upper layer callback that we saved in `D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB` with the help of `this->cb->set_client_cb(cb, dpp);` is invoked in `D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data` in `auto r = client_cb->handle_data(bl, bl_ofs, bl_len);` Now `cb` saved in `D4NFilterObject::D4NFilterReadOp` is invoked, when data is present in the cache and is read from it - this data is returned back to the upper layer/client in `D4NFilterObject::D4NFilterReadOp::flush`, in this line : `int r = client_cb->handle_data(bl, 0, bl.length());`

Reply:
I understand now; thank you for the explanation!